modules/reactor/src/meshgenerators/HexagonMeshTrimmer.C

```diff
@@ -0,0 +1,574 @@
+
+#include "HexagonMeshTrimmer.h"
+#include "MooseMeshUtils.h"
+
+// C++ includes
+#include <cmath> // provides round, not std::round (see http://www.cplusplus.com/reference/cmath/round/)
+
+registerMooseObject("ReactorApp", HexagonMeshTrimmer);
+
+InputParameters
+HexagonMeshTrimmer::validParams()
+{
+  InputParameters params = PolygonMeshGeneratorBase::validParams();
+  params.addRequiredParam<MeshGeneratorName>("input", "The input mesh that needs to be trimmed.");
+  params.addRangeCheckedParam<std::vector<unsigned short>>(
+      "trim_peripheral_region",
+      std::vector<unsigned short>(6, 0),
+      "trim_peripheral_region<=1",
+      "Whether the peripheral region on each of the six sides will be trimmed in an assembly "
+      "mesh. See documentation for numbering convention.");
+  params.addParam<BoundaryName>("peripheral_trimming_section_boundary",
+                                "Boundary formed by peripheral trimming.");
+  params.addRangeCheckedParam<short>(
+      "center_trim_starting_index",
+      "center_trim_starting_index>=0 & center_trim_starting_index<12",
+      "Index of the starting center trimming position.");
+  params.addRangeCheckedParam<short>("center_trim_ending_index",
+                                     "center_trim_ending_index>=0 & center_trim_ending_index<12",
+                                     "Index of the ending center trimming position.");
+  params.addParam<BoundaryName>("center_trimming_section_boundary",
+                                "Boundary formed by center trimming.");
+  params.addParam<BoundaryName>("external_boundary", "External boundary of the input mesh.");
+  params.addParam<SubdomainName>(
+      "tri_elem_subdomain_name_suffix",
+      "trimmer_tri",
+      "Suffix to the block name used for quad elements that are trimmed/converted into "
+      "triangular elements to avert degenerate quad elements");
+  params.addParam<subdomain_id_type>(
+      "tri_elem_subdomain_shift",
+      "Customized id shift to define subdomain ids of the converted triangular elements.");
+
+  params.addClassDescription("This HexagonMeshTrimmer object performs peripheral and/or "
+                             "across-center (0, 0, 0) trimming for "
+                             "assembly or core 2D meshes generated by PatternedHexMG.");
+
+  return params;
+}
+
+HexagonMeshTrimmer::HexagonMeshTrimmer(const InputParameters & parameters)
+  : PolygonMeshGeneratorBase(parameters),
+    _input_name(getParam<MeshGeneratorName>("input")),
+    _trim_peripheral_region(getParam<std::vector<unsigned short>>("trim_peripheral_region")),
+    _peripheral_trimming_section_boundary(
+        isParamValid("peripheral_trimming_section_boundary")
+            ? getParam<BoundaryName>("peripheral_trimming_section_boundary")
+            : BoundaryName()),
+    _trimming_start_sector(isParamValid("center_trim_starting_index")
+                               ? (getParam<short>("center_trim_starting_index") + 9) % 12
```

Comment:
+ 9?
so the convention in the code is different from the convention in the documentation?

Reply:
It is the same. 9 is actually -3 + 12, to solve the same issue that I want -3 % 12 = 9.

Reply:
the % operator being so unintuitive is an issue imo. anyone knows if we have a positive modulo somewhere @lindsayad @roystgnr  ?

Reply:
> the % operator being so unintuitive

Wow, I couldn't agree more.  I always use % to get an unsigned result (content-wise, if not type-wise), so I occasionally do the "add an upper bound multiple of n before taking the modulus" trick to cope with the fact that C/C++ doesn't seem to do Euclidean division properly ... but I'd assumed that the C/C++ behavior must be some standard-among-programmers weirdness ... until I started diving into the rathole of standards and eventually found [this table of modulo operations](https://en.wikipedia.org/wiki/Modulo_operation#In_programming_languages), a total disaster.  Among popular languages only Maple and newer versions of Scheme seem to have first-class support for Euclidean division.  There's an IEEE standard FP remainder operation, but it seems to be the least popular convention in use.  C/C++ `%` on negative numbers was *implementation-defined* until C99 and C++11.  Negative operands on `%` work like C/C++ in many languages, but shader languages leave it undefined, and Lua, Perl, Python, and Ruby all use a different standard.

> anyone knows if we have a positive modulo somewhere

Not in libMesh or dependencies, at least.  I can see why not, too; skimming through `git grep` results shows a whole lot of unsigned modulo operations but I couldn't find any instances where we had to use trickery first.

Reply:
It might be negative as there is some rotation in the algorithm.

Reply:
yes we may have end_index < start_index to get part of the domain. 

let's add a math util ?

Reply:
Sounds like the only self-documenting option to me.  Signature `std::size_t MathUtils::euclidean_mod(T1 dividend, T2 divisor)`?  Fat return type, but shaving it down is probably non-trivial and probably irrelevant for an inlineable function.

`euclidean_mod` is the best name I can think of, but I just noticed that [this reference code for std::div](https://en.cppreference.com/w/cpp/numeric/math/div) calls a function `euclidean_division_string`, shows the example output with negative dividends returning negative remainders, then links right to a wiki page with $0≤r<|b|$ in the first section and negative dividends returning positive remainders in the fourth ... so perhaps that "Euclidean" terminology still isn't as clear as it should be.